Primata/bridge attestors (#21)
movementlabsxyz/MIP: MIP/mip-21/README.md

# MIP-21: Bridge Based on Attestors
- **Description**: Proposal for simplifying the bridge mechanism using an attestor-based approach inspired by USDC's CCTP, reducing friction, cost, and complexity for Layer 2 onboarding.
- **Authors**: [Primata](mailto:primata@movementlabs.xyz)

## Abstract

A bridge mechanism is essential for any Layer 2 solution, serving as the user's first interaction with the network. This process must be smooth and frictionless, ensuring that dealing with tokens on L1 and bridging to native gas tokens on L2 does not overcomplicate the user experience. The bridge must balance user onboarding, cost-efficiency, and simplicity, while minimizing transaction complexity. We propose a bridge based on attestors, inspired by the Cross-Chain Transfer Protocol (CCTP) by USDC, which offers an elegant solution to these challenges.

## Motivation

> Our current bridge solution has been increasing in complexity with each iteration. Originally requiring three on-chain transactions, it now requires five. Additionally, the bridge does not currently support gas charging mechanisms to prevent exploits, and ongoing discussions are leaning toward adding further components, making it even more complex and computationally expensive.

Reviewer: Reference where the solution with 3 and 5 steps is missing

Author: woops its actually 4 onchain transactions:
1. initiate bridge
2. lock bridge
3. complete bridge source chain
4. complete bridge target chain

Reviewer: ok fixed.

We believe the best approach is to draw from live implementations that have achieved a streamlined user experience with minimal cost to the user. One such example is USDC's [CCTP](https://www.circle.com/en/cross-chain-transfer-protocol), which offers simplicity, security, and efficiency in cross-chain bridging. Instead of increasing the complexity of our current bridge design, we should aim to adopt this proven mechanism.

## Specification

The Cross-Chain Transfer Protocol (CCTP) operates using a set of attestors that validate all bridge transactions occurring between chains. These attestors listen for bridge events, confirm the validity of transactions, and, if necessary, revert them by providing appropriate proofs. A transfer allowance managed by a centralized authority further ensures that breaches are mitigated by limiting the amount of value transferred.

To align with this approach and reduce complexity:

> - **Bridge Relayer as Attestor**: The Bridge Relayer in our system will function as an attestor, operating via a multisig. This relayer, through aggregated signatures, produces a proof on a signed message.

Reviewer: ```suggestion
- **Bridge Relayer as Attestor**: The Bridge Relayer in our system will function as an attestor, operating via a multisig scheme (i.e. the attestor collects signatures from a threshold number of comittee members that approve of the message). This relayer, through aggregated signatures, produces a proof on a signed message.
```

  
- **User Interaction**: Users, or other parties interested in completing the bridge, retrieve their funds on the target chain by submitting the signature and message.

> - **Validation Mechanism**: The off-chain signature produced by the multisig is verified against the message on-chain. The challenge is to make sure that the signature is only used once, but the beauty is that the signature is comprised of a multi-party system that do not need to provide onchain signatures. But relying on a protocol like CCTP that has been tested in live environments significantly reduces the risk of any unexpected utilization of off-chain and on-chain proofs.

Reviewer: ```suggestion
- **Validation Mechanism**: The off-chain signature produced by the multisig scheme is verified against the message on-chain. 
The challenge is to make sure that the signature is only used once, but the beauty is that the signature is provided by a multi-party system that do not need to provide onchain signatures. 
Relying on a protocol like CCTP that has been tested in live environments significantly reduces the risk of any unexpected utilization of off-chain and on-chain proofs.
```

Reviewer: Differentiated "multisig" (which is a signature but meant was the scheme) and "the signature".

Or on Circle's own words:

1. An on-chain component on source domain emits a message.
2. Circle's off-chain attestation service signs the message.
3. The on-chain component at the destination domain receives the message, and forwards the message body to the specified recipient.

This approach minimizes the need for additional components and avoids the reliance on synthetic assets, leading to a simplified and cost-effective bridge solution.

## Reference Implementation

For more details on how this model operates, refer to USDC's Cross-Chain Transfer Protocol (CCTP) documentation:  
[USDC CCTP Overview](https://developers.circle.com/stablecoins/docs/generic-message-passing)

## Verification

### 1. **Correctness**: 
The correctness of this proposal can be established by leveraging the success of CCTP, which has been operational and widely adopted.

### 2. **Security Implications**:
The use of a multisig setup adds a layer of security, with aggregated signatures preventing unauthorized access. The transfer allowance mechanism further mitigates potential breaches by limiting how much can be transferred.

### 3. **Performance Impacts**:

> Because we would use a multisig off-chain signature to complete bridges, the set of attestors and the multisig address to produce the signature that we have to check the message against it could be the same set of the L2 Settlement signers.

Reviewer: ```suggestion
Because we would use a multisig off-chain signature to complete bridges, the set of attestors for the bridge and the multisig address to produce the signature that we have to check the message against, could be the same set as the L2 Settlement attestors.
```

Reviewer: Is this what is meant? Are you connecting to our FFS attestor solution?

Author: It could be, not necessarily initially.

Reviewer: Sry, I deleted while you replied. I intended to rephrase. 
.. my deleted question was whether the proposal is that L2 attestors are the same as the attestors described here. 

It's a valid point as L2 settlement attestors are securing safety to some degree already. However we may desire the bridge to have permissioned protection while L2 settlement attestors could be more open. Definitely worth exploring

- **Authors**: [Primata](mailto:primata@movementlabs.xyz)

## Abstract

A bridge mechanism is essential for any Layer 2 solution, serving as the user's first interaction with the network. This process must be smooth and frictionless, ensuring that dealing with tokens on L1 and bridging to native gas tokens on L2 does not overcomplicate the user experience. The bridge must balance user onboarding, cost-efficiency, and simplicity, while minimizing transaction complexity. We propose a bridge based on attestors, inspired by the [Cross-Chain Transfer Protocol](https://developers.circle.com/stablecoins/cctp-getting-started) (CCTP) by USDC, which offers an elegant solution to these challenges.

## Motivation

Our current bridge solution has been increasing in complexity with each iteration. Originally requiring three on-chain transactions, it now requires four, initiate bridge, lock, complete source chain, complete target chain. Additionally, the bridge does not currently support gas charging mechanisms to prevent exploits, and ongoing discussions are leaning toward adding further components, making it even more complex and computationally expensive.

> We believe the best approach is to draw from live implementations that have achieved a streamlined user experience with minimal cost to the user. One such example is USDC's [CCTP](https://www.circle.com/en/cross-chain-transfer-protocol), which offers simplicity, security, and efficiency in cross-chain bridging. Instead of increasing the complexity of our current bridge design, we should aim to adopt this proven mechanism.

Reviewer 2: Circle uses a signing service, [Iris](https://developers.circle.com/stablecoins/generic-message-passing), that current is not decentralized to produce its attestations: https://developers.circle.com/stablecoins/required-block-confirmations

Circle is in effect acting like the trusted relayer in our current architecture, so I don't think this is a good example to draw from under the suggestions of this MIP/MD.

Also, from the docs I've seen, they reference a "native" burn-and-mint process. But, it's unclear why they think this is particularly native.

> We believe the best approach is to draw from live implementations that have achieved a streamlined user experience with minimal cost to the user. One such example is USDC's [CCTP](https://www.circle.com/en/cross-chain-transfer-protocol), which offers simplicity, security, and efficiency in cross-chain bridging. Instead of increasing the complexity of our current bridge design, we should aim to adopt this proven mechanism.

Author: @l-monninger because USDC has bridged version like USDC.e that are ethereum bridged USDC from a distinct bridge. It adds s a second layer of risk. USDC minted by the cross-chain swap (a 1:1 swap) is the same as the one minted by Circle. That's why it's native burn-and-mint.

The thing is, we are over-engineering when a trusted relayer is mandatory. Better to beef up the trusted relayer and rely on it to make it a simpler bridge with less potential exploit surfaces then to over-engineer and open us up to an exploit.

I brought Iris up exactly because we have to trust our relayer just as much as they trust their relayer.

> We believe the best approach is to draw from live implementations that have achieved a streamlined user experience with minimal cost to the user. One such example is USDC's [CCTP](https://www.circle.com/en/cross-chain-transfer-protocol), which offers simplicity, security, and efficiency in cross-chain bridging. Instead of increasing the complexity of our current bridge design, we should aim to adopt this proven mechanism.

Reviewer: i have proposed a split of the scope into 3 sections and reformulated into an MD with the following desiderate
- draw inspiration from CCTP
- attester approach
- specificy the flow

## Specification

The Cross-Chain Transfer Protocol (CCTP) operates using a set of attestors that validate all bridge transactions occurring between chains. These attestors listen for bridge events, confirm the validity of transactions, and, if necessary, revert them by providing appropriate proofs. A transfer allowance managed by a centralized authority further ensures that breaches are mitigated by limiting the amount of value transferred.

To align with this approach and reduce complexity:

> - **Bridge Relayer as Attestor**: The Bridge Relayer in our system will function as an attestor, operating via a multisig scheme (i.e. the attestor collects signatures from a threshold number of comittee members that approve of the message). This relayer, through aggregated signatures, produces a proof on a signed message.  

Author: approvals can either occur onchain or offchain. Multisig for aptos are easier to implement and are better suited for fully onchain, but would very likely lower the rate of processing txs significantly. On ethereum we would use safe.global and their service to propose, sign and complete bridges.

> - **User Interaction**: Users, or other parties interested in completing the bridge, retrieve their funds on the target chain by submitting the signature and message.

Author: bridge goes both ways, but supposing you are talking about eth->move. Currently anyone can impede the user from refunding once the secret is released. On this design anyone can complete the bridge because funds are always sent to the recipient address and the only way of destructuring the transaction is by using the correct recipient address.

The challenge is to make sure that the signature is only used once, but the beauty is that the signature is provided by a multi-party system that do not need to provide onchain signatures.
Relying on a protocol like CCTP that has been tested in live environments significantly reduces the risk of any unexpected utilization of off-chain and on-chain proofs.  

Or on Circle's own words:

1. An on-chain component on source domain emits a message.
2. Circle's off-chain attestation service signs the message.
3. The on-chain component at the destination domain receives the message, and forwards the message body to the specified recipient.

This approach minimizes the need for additional components and avoids the reliance on synthetic assets, leading to a simplified and cost-effective bridge solution.

### Flow

1. **User Deposits**:  
   - The user deposits tokens on the source chain, which triggers a deposit event.
   
2. **Bridge Relayer**:  
   - The Bridge Relayer, listening for the deposit event, picks up the message or messages associated with the deposit.
   

> 3. **Multisig Signing**:  

Reviewer 3: Are you thinking of using Iris for this? I think in general it's fine to draw inspiration from CCTP but more specifics about implementation on Movement would be helpful to get a sense of what is entailed.

Author: No, I think Iris is also closed source.

Reviewer 3: Do you have some other attestation mechanism in mind?

Reviewer: closing this, as 
- the attester may not be the same as FFS.
- this is propossed now as an MD and thus does not require further specification

   - The multisig party signs the bridge transaction. The combination of signatures produces an aggregated signature that ties the signatures to the message.

4. **Target Chain Verification**:  

>    - On the target chain, the aggregated signature and the message are used to verify the validity of the bridge transaction. The withdrawal can only occur if the multisig-provided signature is valid.

Author: I would say that fast finality settlement is not scoped here. But, it can be done so.
First, I advocate for a multisig. This could be 5 keys of hours to prevent the bridge being hacked by stolen keys.
Second, it could be be part of the settlement party.
For the target chain if including FFS, on the layer 2, to learn about the stake, I would not be sure about that, Possibly what we could believe have is a service that adds and removes signers from multisigs on move that correspond to the attestors and their weights (eth is already covered with MCR staking).

>    - On the target chain, the aggregated signature and the message are used to verify the validity of the bridge transaction. The withdrawal can only occur if the multisig-provided signature is valid.

Reviewer: this makes sense. thx for clarifying

5. **User Withdrawal**:  
   - Once the signature is verified, the user can withdraw the corresponding funds on the target chain.

This process is bi-directional, allowing funds to be bridged between Layer 1 and Layer 2, and vice versa, with minimal complexity and friction.

## Reference Implementation

For more details on how this model operates, refer to USDC's Cross-Chain Transfer Protocol (CCTP) documentation:  
[USDC CCTP Overview](https://developers.circle.com/stablecoins/docs/generic-message-passing)

## Verification

### 1. **Correctness**: 

> The correctness of this proposal can be established by leveraging the success of CCTP, which has been operational and widely adopted.

Reviewer 3: Regarding correctness, I don't think any success of CCTP makes sense to use as a reference. 

Nor is CCTP's implementation a reference implementation in an MIP. A reference implementation would mean an implementation for Movement's bridge.

I would change this so it either consistently mentions CCTP as an implementation from which this MIP draws inspiration, or include some code as a reference implementation for Movement's bridge. The key details have to do with how the Solidity modules and Move contracts would need to be modified, along with the bridge infrastructure.

Author: Not sure if I understood, I thik that this would be a major overhaul. But at least we would bootstrap our security with a different model. Major work would be the logic on the move modules.

Reviewer: I support @andygolay in his statement. I think the MIP in its current form would classify as an MD, unless further details go into this about how specifically the components would be implemented. 

@l-monninger you are the MD master, is this content what you had in mind for an MD?

Reviewer: @0xPrimata I have moved it over to MD type. are you ok with this change?
